File: go/socketd/transport/core/impl/channel_base.go
package impl

import (
	"net/url"
	"strings"

	"github.com/google/uuid"
	"socketd/transport/core"
	"socketd/transport/core/constant"
	"socketd/transport/core/message"
)

//var _ core.Channel = new(ChannelBase)

type ChannelBase struct {
	core.Channel

	config      core.Config
	attachments map[string]any
	handshake   *core.Handshake
	isClosed    int
}

func NewChannelBase(config core.Config) *ChannelBase {
	return &ChannelBase{config: config}
}

func (c *ChannelBase) GetConfig() core.Config {
	return c.config
}

func (c *ChannelBase) GetAttachment(name string) any {
	val, ok := c.attachments[name]
	if ok {
		return val
	}
	return nil
}

func (c *ChannelBase) PutAttachment(name string, val any) {
	if val == nil {
		delete(c.attachments, name)
		return
	}
	c.attachments[name] = val
}

func (c *ChannelBase) IsClosed() int {
	return c.isClosed
}

func (c *ChannelBase) Close(code int) {
	c.isClosed = code
	//for k := range c.attachments {
	//	delete(c.attachments, k)
	//}
}

func (c *ChannelBase) SetHandshake(handshake *core.Handshake) {
	c.handshake = handshake
}

func (c *ChannelBase) GetHandshake() *core.Handshake {
	return c.handshake
}

func (c *ChannelBase) SendConnect(event string, metas map[string]string) (err error) {
	sid := strings.Replace(uuid.NewString(), "-", "", -1)
	var uv = url.Values{}
	for k, v := range metas {
		uv.Set(k, v)
	}
	frame := message.NewFrame(constant.FrameConnect, message.NewMessage(sid, event, message.NewEntity(uv, nil)))
	return c.Send(frame, nil)
}

func (c *ChannelBase) SendConnectAck(connectMsg *message.Message) (err error) {
	var frame = &message.Frame{}
	frame.Flag = constant.FrameConnack
	frame.Message = connectMsg
	return c.Send(frame, nil)
}

func (c *ChannelBase) SendPing() (err error) {
	var frame = &message.Frame{}
	frame.Flag = constant.FramePing
	return c.Send(frame, nil)
}

func (c *ChannelBase) SendPong() (err error) {
	var frame = &message.Frame{}
	frame.Flag = constant.FramePong
	return c.Send(frame, nil)
}

func (c *ChannelBase) SendClose() (err error) {
	var frame = &message.Frame{}
	frame.Flag = constant.FrameClose
	return c.Send(frame, nil)
}

func (c *ChannelBase) SendAlarm(from *message.Message, alarm string) (err error) {
	// TODO panic("implement me")
	panic("implement me")
	return nil
}

File: java/socketd/src/main/java/org/noear/socketd/transport/core/impl/ChannelBase.java
package org.noear.socketd.transport.core.impl;

import org.noear.socketd.transport.core.*;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 通道基类
 *
 * @author noear
 * @since 2.0
 */
public abstract class ChannelBase implements Channel {
    //最大请求数（根据请求、响应加减计数）
    private final Config config;
    //附件
    private final Map<String, Object> attachments = new ConcurrentHashMap<>();
    //握手信息
    private HandshakeInternal handshake;


    public ChannelBase(Config config) {
        this.config = config;
    }


    @Override
    public Config getConfig() {
        return config;
    }

    @Override
    public <T> T getAttachment(String name) {
        return (T) attachments.get(name);
    }

    @Override
    public void putAttachment(String name, Object val) {
        if (val == null) {
            attachments.remove(name);
        } else {
            attachments.put(name, val);
        }
    }

    @Override
    public void setHandshake(HandshakeInternal handshake) {
        if(handshake != null) {
            this.handshake = handshake;
        }
    }


    @Override
    public HandshakeInternal getHandshake() {
        return handshake;
    }

    @Override
    public void sendConnect(String uri, Map<String, String> metaMap) throws IOException {
        send(Frames.connectFrame(getConfig().genId(), uri, metaMap), null);
    }

    @Override
    public void sendConnack() throws IOException {
        send(Frames.connackFrame(getHandshake()), null);
    }

    @Override
    public void sendPing() throws IOException {
        send(Frames.pingFrame(), null);
    }

    @Override
    public void sendPong() throws IOException {
        send(Frames.pongFrame(), null);
    }

    @Override
    public void sendClose(int code) throws IOException {
        send(Frames.closeFrame(code), null);
    }

    @Override
    public void sendAlarm(Message from, Entity alarm) throws IOException {
        send(Frames.alarmFrame(from, alarm), null);
    }

    @Override
    public void sendPressure(Message from, Entity pressure) throws IOException {
        send(Frames.pressureFrame(from, pressure), null);
    }

    @Override
    public void close(int code) {
        if (code > Constants.CLOSE1000_PROTOCOL_CLOSE_STARTING) {
            attachments.clear();
        }
    }
}
